File: LeetCode/Medium/7 ReverseInteger/7ReverseInteger.cpp
#include <iostream>
#include <climits>  // para INT_MAX e INT_MIN

using namespace std;

class Solution {
public:
    int reverse(int x) {
        int reversed = 0;
        while (x != 0) {
            int digit = x % 10;
            if (reversed > INT_MAX / 10 || reversed < INT_MIN / 10) {
                return 0;
            }
            reversed = reversed * 10 + digit;
            x /= 10;
        }
        return reversed;
    }
};

int main() {
    Solution sol;
    cout << sol.reverse(123) << endl;
    cout << sol.reverse(456) << endl;
    cout << sol.reverse(-123) << endl;
    cout << sol.reverse(1534236469) << endl;

    return 0;
}

File: LeetCode/Medium/7 ReverseInteger/7ReverseInteger.py
def reverse(x: int) -> int:
    INT_MAX = 2**31 - 1   # 2147483647
    INT_MIN = -2**31      # -2147483648
    sign = -1 if x < 0 else 1
    x = abs(x)

    reversed_num = 0
    while x != 0:
        digit = x % 10
        if reversed_num > INT_MAX // 10:
            return 0

        reversed_num = reversed_num * 10 + digit
        x //= 10

    reversed_num *= sign

    # Verificacion final de rango (por el signo aplicado al final):
    if reversed_num < INT_MIN or reversed_num > INT_MAX:
        return 0

    return reversed_num

def main():
    print(reverse(123))
    print(reverse(456))
    print(reverse(-123))
    print(reverse(1534236469)) 

if __name__ == "__main__":
    main()